common/codex32.c

```diff
@@ -0,0 +1,170 @@
+#include "config.h"
+#include <assert.h>
+#include <bitcoin/chainparams.h>
+#include <ccan/tal/str/str.h>
+#include <common/bech32.h>
+#include <common/bech32_util.h>
+#include <common/bolt12.h>
+#include <common/bolt12_merkle.h>
+#include <common/codex32.h>
+#include <common/configdir.h>
+#include <common/features.h>
+#include <math.h>
+#include <secp256k1_schnorrsig.h>
+#include <time.h>
+
+
+/* Helper to sanity check the codex32 string parts */
+static bool sanity_check (struct codex32_parts *parts,
+			  char **fail)
+{
+	if (!streq(parts->hrp, "ms") && !streq(parts->hrp, "MS")) {
+		*fail = tal_fmt(tmpctx, "Invalid HRP!");
+		return false;
+	}
+	if (parts->threshold > 9 ||
+	    parts->threshold < 0 ||
+	    parts->threshold == 1) {
+		*fail = tal_fmt(tmpctx, "Invalid threshold!");
+		return false;
+	}
+	if (strlen(parts->id) != 4) {
+		*fail = tal_fmt(tmpctx, "Invalid ID!");
+		return false;
+	}
+	if ((parts->threshold == 0 && !(*(parts->share_idx) == 'S' || *(parts->share_idx) == 's')))
+	{
+		*fail = tal_fmt(tmpctx, "Expected share index S for threshold 0!");
+		return false;
+	}
+	if((strlen(parts->payload) * 5) % 8 > 4) {
+		*fail = tal_fmt(tmpctx, "Incomplete group exist in payload!");
+		return false;
+	}
+
+	return true;
+}
+
+static const u8 *extract_data(struct codex32_parts *parts)
+{
+	// FIXME: Make sure the size of array is correct, because the documentation has 1 extra byte...
+	u8 *ret = tal_arr(parts, u8, ((strlen(parts->payload) * 5 + 7) / 8) - 1);
+
+	uint8_t next_byte = 0;
+	uint8_t rem = 0;
+	size_t i = 0, j = 0;
+	while (parts->payload[i] != '\0') {
+		char ch = parts->payload[i++];
+		uint8_t fe = bech32_charset_rev[(int)ch];
+
+		if (rem < 3) {
+			// If we are within 3 bits of the start we can fit the whole next char in
+			next_byte |= fe << (3 - rem);
+		}
+		else if (rem == 3) {
+			// If we are exactly 3 bits from the start then this char fills in the byte
+			ret[j++] = next_byte | fe;
+			next_byte = 0;
+		}
+		else { // rem > 3
+			// Otherwise we have to break it in two
+			uint8_t overshoot = rem - 3;
+			assert(overshoot > 0);
```

Comment:
int overshoot would allow a slightly stronger check here?  And uint8_t we usually simply call u8, using ccan/short_types/short_types.h.